torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,195 @@
+import itertools
+from typing_extensions import Protocol
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class _LazyProtocol(Protocol):
+    """This is to avoid errors with mypy checks for 
+    The attributes in a mixin:
+    https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes
+    """
+    def _register_load_state_dict_pre_hook(self, hook):
+        ...
+
+    def register_forward_pre_hook(self, hook):
+        ...
+
+    def _lazy_load_hook(
+            self, state_dict, prefix, local_metadata, strict,
+            missing_keys, unexpected_keys, error_msgs):
+        ...
+
+    def _get_name(self):
+        ...
+
+    def _infer_parameters(self, module, input):
+        ...
+
+    @property
+    def _parameters(self):
+        ...
+
+    @property
+    def _buffers(self):
+        ...
+
+    @property
+    def _non_persistent_buffers_set(self):
+        ...
+
+    @property
+    def _load_hook(self):
+        ...
+
+    @property
+    def _initialize_hook(self):
+        ...
+
+
+class LazyModuleMixin:
+    r"""A mixin for modules that lazily initialize parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" ``forward`` once to initialize all parameters in case
+    that you need to access the weights before the actual "real" ``forward`` pass.
+
+    >>> lazy_module = torch.nn.LazyLinear(10)
+    >>> lazy_module.forward(input)  # Dummy forward call to initialize the module parameters
+    >>> assert type(lazy_module) is torch.nn.Linear
+
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+
+    The default value of lazy parameters are ``UninitializedParameter()``.
+
+    .. note:: A `LazyModule`, or a module containing a `LazyModule` requires an explicit
```

Comment:
Maybe we should just make the initial dry run part of the lazy pattern? This will be even more important if lazy modules aren't serializable. Also, if we think about PyTorch networks being modular then it's hard to know how someone else might use them, making an initial dry run an important part of the program.

Reply:
lazy modules are currently serializable. 
What do you mean by making the initial dry run part of the lazy pattern? 
adding some kind of explicit `initialize` method to a lazy module?

Reply:
> lazy modules are currently serializable.

I understand. But should they be? I'm not sure I understand the benefit.

> What do you mean by making the initial dry run part of the lazy pattern?
> adding some kind of explicit `initialize` method to a lazy module?

Not exactly. I mean that we tell people the way to use modules is to put the network on the appropriate device and dtype, perform a dry run, then continue modifying the network or accessing its parameters (by, for example, attaching an optimizer to it). Anything else could throw an error message.

Reply:
Yes, more or less this is how it currently works, but the dry_run is only mandatory in certain cases. If I understood this correctly, we should make the dry run mandatory regardless of the use case. It is a change that I agree with as it makes everything explicit.

Reply:
You're completely correct. However, following our other discusisons, for now let's remove this note. In the future we can be more explicit about when a dry run isn't required.